### PR TITLE
8264954: unified handling for VectorMask object re-materialization during de-optimization

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Following flow describes object reconstruction for de-optimization:-
1)	PhaseVector::scalarize_vbox_node() creates SafePointScalarObjectNode to captures the box type information, also it connects to node holding the boxed value.
2)	During code emit phase (PhaseOutput) C2 process above information to dumps ObjectValue holding the box information and LocationValue to holding the value information into ScopeDescriptor corresponding to Safepoint PC.
3)	De-optimization blobs dump the value held in registers to the stack locations using RegisterSave::save_live_registers() and a mapping b/w register and its stack location is added to RegisterMap.
4)	During de-optimization, compiled frame objects are re-allocated using identity information held in ObjectValue and their fields are initialized using values held in the stack locations accessed through register-stack mappings. 

By inserting a VectorStoreMaskNode before stitching the mask holding node to Safepoint we make sure that value held in opmask/vector register is transferred to a byte vector. Thus rest of the flow works as it is, stack location will hold the value in the form of a byte array irrespective of the box shape.

tier1-tier3 regressions are clean with UseAVX=2/3.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264954](https://bugs.openjdk.java.net/browse/JDK-8264954): unified handling for VectorMask object re-materialization during de-optimization

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 239923fd891a84545b7b6a0c71355d2927e7f827

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3408/head:pull/3408` \
`$ git checkout pull/3408`

Update a local copy of the PR: \
`$ git checkout pull/3408` \
`$ git pull https://git.openjdk.java.net/jdk pull/3408/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3408`

View PR using the GUI difftool: \
`$ git pr show -t 3408`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3408.diff">https://git.openjdk.java.net/jdk/pull/3408.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/vector.cpp [hunk 1]
```diff
@@ -252,6 +261,22 @@ void PhaseVector::scalarize_vbox_node(VectorBoxNode* vec_box) {
...
+    bool is_mask = is_vector_mask(iklass);
+    if (is_mask) {
+      if (vec_value->Opcode() == Op_VectorLoadMask) {
+        const Type* in1_ty = vec_value->in(1)->bottom_type();
+        assert(in1_ty->isa_vect() && in1_ty->is_vect()->element_basic_type() == T_BOOLEAN, "");
+        vec_value = vec_value->in(1);
+      } else if (vec_value->Opcode() != Op_VectorStoreMask) {
+        const TypeVect* vt = vec_value->bottom_type()->is_vect();
+        BasicType bt = vt->element_basic_type();
+        vec_value = gvn.transform(VectorStoreMaskNode::make(gvn, vec_value, bt, vt->length()));
```
Comment (iwanowww) [inline src/hotspot/share/opto/vector.cpp:277]: `VectorStoreMaskNode::Identity()` already handles `VectorStoreMask (VectorLoadMask bv) elem_size ==> bv` transformation, doesn't it? So, you can just unconditionally instantiate `VectorStoreMaskNode` and let IGVN handle it.

Also, an observation on naming: `VectorLoadMask` and `VectorStoreMask` names are misleading. On the surface, they look like memory nodes, but in fact, are cast nodes (`Vector <-> Mask`). Please, file an RFE to address that eventually.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 2]
```diff
@@ -88,20 +88,13 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
 
 void VectorSupport::init_payload_element(typeArrayOop arr, bool is_mask, BasicType elem_bt, int index, address addr) {
   if (is_mask) {
-    // Masks require special handling: when boxed they are packed and stored in boolean
```
Comment (iwanowww) [inline src/hotspot/share/prims/vectorSupport.cpp:91]: The comment is still valid. Please, keep it and update with latest details about mask support.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 3]
```diff
@@ -88,20 +88,13 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
...
-      case T_INT:    // fall-through
-      case T_FLOAT:  arr->bool_at_put(index,   (*(jint*)addr) != 0); break;
-      case T_LONG:   // fall-through
-      case T_DOUBLE: arr->bool_at_put(index,  (*(jlong*)addr) != 0); break;
+      case T_BYTE:
+      case T_SHORT:
+      case T_INT:
+      case T_LONG:
+      case T_FLOAT:
+      case T_DOUBLE: arr->bool_at_put(index,  (*(jbyte*)addr) != 0); break;
```
Comment (iwanowww) [inline src/hotspot/share/prims/vectorSupport.cpp:97]: No switch needed anymore. Just leave the  `arr->bool_at_put(index,  (*(jbyte*)addr) != 0)`.
As an option, consider adding an assert (`assert(is_java_type(bt) && bt != T_BOOLEAN)`.

Comment (iwanowww) [reply src/hotspot/share/prims/vectorSupport.cpp:97]: Actually, I think you can get rid of `is_mask` and just use `elem_bt = T_BOOLEAN` / `bool_at_put(index, (*(jboolean*)addr))`.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 4]
```diff
@@ -124,7 +117,11 @@ Handle VectorSupport::allocate_vector_payload_helper(InstanceKlass* ik, frame* f
 
   int num_elem = klass2length(ik);
   BasicType elem_bt = klass2bt(ik);
-  int elem_size = type2aelembytes(elem_bt);
+  // Inserting a VectorStoreMask before stitching the mask
+  // to SafePointNode will ensure packing the mask into a
+  // byte array for masks present in both predicated register
+  // or vector registers.
+  int elem_size = is_mask ? 1 : type2aelembytes(elem_bt);
```
Comment (iwanowww) [inline src/hotspot/share/prims/vectorSupport.cpp:124]: If `elem_bt == T_BOOLEAN` for masks, you can get rid of `is_mask` usages in `VectorSupport::allocate_vector_payload_helper()` and just introduce a special case in `VectorSupport::klass2bt` (as already done for `VectorShuffle`s).

**Review by iwanowww — COMMENTED** [review]
Very nice! I like your idea to unify mask representation. It simplifies implementation a lot.
But it also means there'll be 2 representations kept alive for masks with debug usages and burns 2 registers (speaking of x86, either 1 predicate + 1 vector registers on AVX512 or 2 vector regsiters on pre-AVX512). Do you see any problems with that? (It can be improved later if it turns out to be a problem in practice.)

**Comment by jatin-bhateja** [comment]
Thanks @iwanowww , your comments are resolved.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 5]
```diff
@@ -86,48 +88,36 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
...
-      default: fatal("unsupported: %s", type2name(elem_bt));
-    }
+// Masks require special handling: when boxed they are packed and stored in boolean
+// arrays, but in scalarized form they have the same size as corresponding vectors.
+// For example, Int512Mask is represented in memory as boolean[16], but
+// occupies the whole 512-bit vector register when scalarized.
+// During scalarization inserting a VectorStoreMask node between mask
+// and safepoint node always ensures the existence of masks in a boolean array.
+//
+// TODO: revisit when predicate registers are fully supported.
```
Comment (iwanowww) [inline src/hotspot/share/prims/vectorSupport.cpp:98]: TODO line can be removed now.

File: src/hotspot/share/opto/vector.cpp [hunk 6]
```diff
@@ -252,6 +261,18 @@ void PhaseVector::scalarize_vbox_node(VectorBoxNode* vec_box) {
 #endif // ASSERT
                                                first_ind, n_fields);
     sobj->init_req(0, C->root());
+
+    // If a mask is feeding into a safepoint, then its value should be
+    // packed into a byte vector first, this will simplify the re-materialization
+    // logic for both predicated and non-predicated targets.
+    bool is_mask = is_vector_mask(iklass);
+    if (is_mask) {
+      if (vec_value->Opcode() != Op_VectorStoreMask) {
```
Comment (iwanowww) [inline src/hotspot/share/opto/vector.cpp:270]: It's better to do the check against the type of `vec_value`: `VectorStoreMask` produces a vector of booleans  and you can guard `gvn.transform(VectorStoreMaskNode::make(...))` call with `bt != T_BOOLEAN` check instead.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 7]
```diff
@@ -86,48 +88,36 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
...
+//
+void VectorSupport::init_payload_element(typeArrayOop arr, BasicType elem_bt, int index, address addr) {
+  switch (elem_bt) {
+    case T_BOOLEAN:
```
Comment (iwanowww) [inline src/hotspot/share/prims/vectorSupport.cpp:102]: Please, use `bool_at_put()` accessor here (just for clarity):
```
case T_BOOLEAN: arr->bool_at_put(index, *(jboolean*)addr); break;
```

**Review by iwanowww — APPROVED** [review]
Looks good. 

Minor comments follow.

File: src/hotspot/share/opto/vector.cpp (hunk 6 above)
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vector.cpp:270]: In addition to bt != T_BOOLEAN check, we also need to check if type is a vectormask (explicit type added for mask generating nodes on targets supporting predicate registers).

A similar check needs to be added at following location
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vector.cpp#L329

Keeping this as it is i.e. an Opcode based check.

Comment (iwanowww) [reply src/hotspot/share/opto/vector.cpp]: > In addition to bt != T_BOOLEAN check, we also need to check if type is a vectormask (explicit type added for mask generating nodes on targets supporting predicate registers).

You check here that the mask is in canonical representation (which is a vector of booleans) before applying the canonicalizing conversion. Alternatively, you could check that the mask is in "native"/platform-dependent representation (`vec_value` has `TypeVectMask` type).

So, my reading of your response is that `bt != T_BOOLEAN` check is not enough to reliably distinguish between the representations: a mask for vector of booleans (in native representation) will have the same type as the vector of booleans produced by the canonicalizing cast of a mask value. I agree with that.

But until `TypeVectMask` is used uniformly and independently of predicate register support, `TypeVect` vs `TypeVectMask` doesn't cut the problem as well. And adding `bt != T_BOOLEAN` doesn't help.

So, I'm fine with leaving the Opcode check for now. 

In the longer term, we need to come up with a uniform representation in type system for masks in native format to be able to reliably distinguish between masks and vectors in cross-platform manner.

**Review by dcubed-ojdk — COMMENTED** [review]
<Remove comment posted in error>

**Pull request closed without merging**